lib/src/http.dart

```diff
@@ -343,58 +231,202 @@ Map parseJsonResponse(http.Response response) {
+        // is tempting to handle these error codes before retrying, saving time
+        // for the end-user, it is known that DNS lookups can fail intermittently
+        // in some cloud environments. Furthermore, since these error codes are
+        // platform-specific (undocumented) and essentially cargo-culted along
+        // skipping retries may lead to intermittent issues that could be fixed
+        // with a retry. Failing to retry intermittent issues is likely to cause
+        // customers to wrap pub in a retry loop which will not improve the
+        // end-user experience.
+
+        // TODO: should these return PubConnectException('msg', couldRetry: false)?
+        if (osError.isDnsError) {
+          fail('Could not resolve URL "${request.url.origin}".');
+        } else if (osError.isSslError) {
+          fail(
+              'Unable to validate SSL certificate for "${request.url.origin}".');
+        }
+      }
+
+      return e;
+    }, retryIf: (e, stackTrace) async {
+      // TODO: think about IOException. RetryClient used to catch IOException
+      // from _PubHttpClient.
+
+      if (e is PubConnectException && e.couldRetry) return true;
+      if (e is PubHttpException && e.couldRetry) return true;
+
+      return (e is PubHttpException && e.couldRetry) ||
+          e is HttpException ||
+          e is TimeoutException ||
+          e is FormatException ||
+          (retryIf != null && await retryIf(e, stackTrace));
+    }, onRetry: (exception, retryCount) async {
+      if (onRetry != null) {
+        await onRetry(exception, retryCount, request, managedResponse);
+      } else {
+        log.io('Retry #${retryCount + 1} for '
+            '${request.method} ${request.url}...');
+      }
+
+      if (retryCount != 3) return;
+      if (!_retriedHosts.add(request.url.host)) return;
+      log.message('It looks like ${request.url.host} is having some trouble.\n'
+          'Pub will wait for a while before trying to connect again.');
+    }, maxAttempts: maxAttempts);
+  }
+}
+
+extension on http.BaseResponse {
+  void throwIfError() {
+    // Retry if the response indicates a server error.
+    if ([500, 502, 503, 504].contains(statusCode)) {
+      throw PubHttpException(this as http.Response, couldRetry: true);
+    }
+
+    // 401 responses should be handled by the OAuth2 client. It's very
+    // unlikely that they'll be returned by non-OAuth2 requests. We also want
+    // to pass along 400 responses from the token endpoint.
+    var tokenRequest = request!.url == oauth2.tokenEndpoint;
+    if (statusCode < 400 ||
+        statusCode == 401 ||
+        (statusCode == 400 && tokenRequest)) {
+      return;
+    }
+
+    if (statusCode == 406 &&
+        request!.headers['Accept'] == pubApiHeaders['Accept']) {
+      fail('Pub ${sdk.version} is incompatible with the current version of '
+          '${request!.url.host}.\n'
+          'Upgrade pub to the latest version and try again.');
+    }
```

Comment:
I don't think these can be null, but this is an error message, if we get it -- it's better that we get null than an null-error :see_no_evil: 
```suggestion
    if (statusCode == 406 &&
        request?.headers['Accept'] == pubApiHeaders['Accept']) {
      fail('Pub ${sdk.version} is incompatible with the current version of '
          '${request?.url.host}.\n'
          'Upgrade pub to the latest version and try again.');
    }
```